src/metatrain/experimental/soap_bpnn/trainer.py

```diff
@@ -46,18 +49,42 @@ def train(
         validation_datasets: List[Union[Dataset, torch.utils.data.Subset]],
         checkpoint_dir: str,
     ):
-        dtype = train_datasets[0][0]["system"].positions.dtype
 
-        # only one device, as we don't support multi-gpu for now
-        assert len(devices) == 1
-        device = devices[0]
+        is_distributed = self.hypers["distributed"]
+
+        if is_distributed:
+            distr_env = DistributedEnvironment(self.hypers["distributed_port"])
+            torch.distributed.init_process_group(backend="nccl")
+            world_size = torch.distributed.get_world_size()
+            rank = torch.distributed.get_rank()
+        else:
+            rank = 0
+
+        if is_distributed:
+            if len(devices) > 1:
+                raise ValueError(
+                    "Requested distributed training with the `multi-gpu` device. "
+                    " If you want to run distributed training with SOAP-BPNN, please "
+                    "set `device` to cuda."
+                )
+            device = torch.device("cuda", distr_env.local_rank)
```

Comment:
This seems to be a potential issue. For example, setting the device to `cuda:<local_rank>` works on izar, where the GPUs on a given node are not hidden from one another, but crashes on CSCS, where the GPUs are hidden and `cuda:0` is correct instead

Reply:
The way I handled this in lammps was to use `distr_env.local_rank % torch.cuda.device_count()`. This will handle the case of hidden devices (since `torch.cuda.device_count()` only returns visible devices), and the case where using more ranks than available GPUs.

Reply:
This sounds like a good idea, thanks a lot